functional/lenses.md

```diff
@@ -0,0 +1,481 @@
+
+```json
+{ "primary_customer_id": 1048576332,
+  "accounts"; [
+      { "account_id": 2121,
+        "account_type: "savings",
+        "joint_customer_ids": [],
+        [...]
+      },
+      { "account_id": 2122,
+        "account_type: "checking",
+        "joint_customer_ids": [1048576333],
+        [...]
+      },
+  ]
+} 
+```
+
+It would be possible to create a `struct` for each of these types and parse the
+JSON into it.
+However, each one would need to implement their own version of, for example,
+`get_customer_id`.
+But what if one piece of code wanted to retrieve all customer IDs regardless of
+data type?
+
+The solution in Rust is to make a trait which represents the operation.
+If it is implemented for both `struct`s, then type erasure can make the once
+piece of code work:
+
+```rust
+use std::collections::HashSet;
+
+pub trait CustomerId {
+    fn get_customer_id(&self) -> u64;
+}
+
+pub struct CreditRecord {
+    customer_id: u64,
+    name: String,
+    dob: String,
+    // other fields omitted
+}
+
+pub struct Account {
+    account_id: u32,
+    account_type: String,
+    // other fields omitted
+}
+
+pub struct AccountRecord {
+    primary_customer_id: u64,
+    accounts: Vec<Account>,
+}
+
+impl CustomerId for CreditRecord {
+    fn get_customer_id(&self) -> u64 {
+        self.customer_id
+    }
+}
+
+impl CustomerId for AccountRecord {
+    fn get_customer_id(&self) -> u64 {
+        self.primary_customer_id
+    }
+}
+
+// static polymorphism: allows one singular type at a time, but a function call
+// with any type
+fn unique_ids_set<R: CustomerId>(records: &[R]) -> HashSet<u64> {
+    records.iter().map(|r| r.get_customer_id()).collect()
+}
+
+// dynamic polymorphism: iterates over any type, allowing collecting different
+// records together
+fn unique_ids_boxed<I>(iterator: I) -> HashSet<u64>
+    where I: Iterator<Item=Box<dyn CustomerId>>
+{
+    iterator.map(|r| r.as_ref().get_customer_id()).collect()
+}
+```
+
+The versions of `unique_ids` shown here implement the *concept* of a lens.
+
+Lenses are a way to allow accessing parts of a data type in an abstract,
+unified way.[^1]
+Lenses define a generic operation -- in our case, getting a customer ID from a
+record -- that can apply to multiple types.
+This is so even though each type has a different way to "observe" the desired
+data.
+
+At first glance, it may seem more like Rust is using "duck typing", the way
+that dynamic languages do.
+However, there are significant differences:
+
+1. Static type checking is maintained. Only types which "opt-in" to having a
+   customer ID (with an `impl CustomerId` block) can be placed into the
+   iterator or collection. Duck typed languages require this to be a runtime
+   check, which can then go wrong at runtime.
+1. The `impl CustomerId` block lives with the type. This means the type can
+   live in a different crate than definition of `CustomerId` itself,
+   maintaining extensibility.
```

Comment:
You may want to rephrase this a bit. Impls can also reside in the same module as the trait definition. _I_ get what you are saying, but people new to Rust may get confused a little by this statement. Maybe add something like `in this case` or `in this example`.